### PR TITLE
[inductor] Improve merging of contiguous loops

### PR DESCRIPTION
**Pull request by jansel**

### CODE REVIEW

File: torchinductor/dependencies.py [hunk 1]
```diff
@@ -71,6 +80,10 @@ def store(self, name, index, value):
     def reduction(self, name, dtype, reduction_type, index, value):
         return self.store(name, index, f"reduce_{reduction_type})({value})")
 
+    def index_expr(self, expr, dtype):
```
Comment (desertfire) [inline torchinductor/dependencies.py:83]: Curious, where is this called?

Comment (jansel) [reply torchinductor/dependencies.py:83]: `torch.arange`, `torch.triu`, padding, etc.  It is when there is data flow from indexing variables into compute.

File: torchinductor/ir.py [hunk 2]
```diff
@@ -42,6 +60,15 @@ def eval(cls, base, divisor, modulus):
             and isinstance(modulus, sympy.Integer)
         ):
             return (base // divisor) % modulus
+
+        if divisor == 1:
+            w1 = sympy.Wild("w1")
+            w2 = sympy.Wild("w2")
+            m = base.match(w1 + modulus * w2)
+            if m and "/" not in str(m[w1]) and "/" not in str(m[w2]) and m[w2] != 0:
```
[the remark this answers is not in the document]
Comment (jansel) [reply torchinductor/ir.py:69]: done

File: torchinductor/ir.py [hunk 3]
```diff
@@ -1060,6 +1123,118 @@ def decide_layout(self):
...
+                    read_writes.writes,
+                    read_writes.index_exprs,
+                )
+            }
+        )
+        memory_addrs = [
+            dep.index for dep in itertools.chain(read_writes.reads, read_writes.writes)
+        ]
+
+        def simplify_and_reorder(vars_, sizes):
```
Comment (desertfire) [inline torchinductor/ir.py:1145]: Do we have a flag to control this? At least for the reordering part? I am thinking about future auto-tuning.

**Review by desertfire — APPROVED** [review]
Nervously approve without understanding all the details.

File: torchinductor/ir.py (hunk 3 above)
Comment (jansel) [reply torchinductor/ir.py:1145]: `torchinductor.config.pick_loop_orders` turns on/off the reordering.

In the future we should definitely autotune the actual orders, or use some sort of global solver-based system.

**Pull request merged**